### PR TITLE
[FnsOverBlob][Internal Repo] Add tests for create implicit and explicit

### PR DESCRIPTION
**Pull request by anmolanmol1234**

### CODE REVIEW

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java [hunk 1]
```diff
@@ -677,7 +677,7 @@ public abstract AbfsRestOperation flush(byte[] buffer,
       final String leaseId,
       final String eTag,
       ContextEncryptionAdapter contextEncryptionAdapter,
-      final TracingContext tracingContext) throws AzureBlobFileSystemException;
+      final TracingContext tracingContext, final String md5Hash) throws AzureBlobFileSystemException;
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java]: Do we need to change the method signature?
Can we compute Md5 inside BlobClient Itself?

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java [hunk 2]
```diff
@@ -763,9 +764,11 @@ public synchronized void close() throws IOException {
               tracingHeaderFormat,
               listener, abfsCounters.toString());
       try {
-        getAbfsClient().getMetricCall(tracingMetricContext);
+        if (!tracingMetricContext.getMetricResults().equals(EMPTY_STRING)) {
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java]: Nit: Can be simplified for OSS as follow?
```
if (abfsStore.getClient().isMetricCollectionEnabled() && !abfsCounters.toString().equals(EMPTY_STRING)) {
      try {
        TracingContext tracingMetricContext = new TracingContext(
            clientCorrelationId,
            fileSystemId, FSOperationType.GET_ATTR, true,
            tracingHeaderFormat,
            listener, abfsCounters.toString());
        getAbfsClient().getMetricCall(tracingMetricContext);
      } catch (IOException e) {
        LOG.error("Error while getting metrics from client", e);
      }
 }
```

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 3]
```diff
@@ -966,29 +971,31 @@ AZURE_FOOTER_READ_BUFFER_SIZE, getAbfsConfiguration().getFooterReadBufferSize())
...
       AbfsClient writeClient = getClientHandler().getIngressClient();
-
       final AbfsRestOperation op = getClient()
           .getPathStatus(relativePath, false, tracingContext, null);
       perfInfo.registerResult(op.getResult());
 
-      final String resourceType = getClient().checkIsDir(op.getResult()) ? DIRECTORY : FILE;
-      final Long contentLength = Long.valueOf(op.getResult().getResponseHeader(HttpHeaderConfigurations.CONTENT_LENGTH));
-
+      final String resourceType = getClient().checkIsDir(op.getResult())
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: Nit: Can be simplified for OSS as follows?? We don't need resourceType variable
```
      final AbfsRestOperation op = getClient()
          .getPathStatus(relativePath, false, tracingContext, null);
      perfInfo.registerResult(op.getResult());
      
      if (getClient().checkIsDir(op.getResult())) {
        throw new AbfsRestOperationException(
            AzureServiceErrorCode.PATH_NOT_FOUND.getStatusCode(),
            AzureServiceErrorCode.PATH_NOT_FOUND.getErrorCode(),
            "openFileForWrite must be used with files and not directories",
            null);
      }
```

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 4]
```diff
@@ -896,8 +897,12 @@ public AbfsInputStream openFileForRead(Path path,
         AbfsHttpOperation op = getClient().getPathStatus(relativePath, false,
             tracingContext, null).getResult();
         resourceType = getClient().checkIsDir(op) ? DIRECTORY : FILE;
-        contentLength = Long.parseLong(
-            op.getResponseHeader(HttpHeaderConfigurations.CONTENT_LENGTH));
+        String contentLengthHeader = op.getResponseHeader(HttpHeaderConfigurations.CONTENT_LENGTH);
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: We can create a utility method extractContentLength(op.getResult()) having these checks.
Same can be used at other places as well.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 5]
```diff
@@ -966,29 +971,31 @@ AZURE_FOOTER_READ_BUFFER_SIZE, getAbfsConfiguration().getFooterReadBufferSize())
...
-                AzureServiceErrorCode.PATH_NOT_FOUND.getStatusCode(),
-                AzureServiceErrorCode.PATH_NOT_FOUND.getErrorCode(),
-                "openFileForWrite must be used with files and not directories",
-                null);
+            AzureServiceErrorCode.PATH_NOT_FOUND.getStatusCode(),
+            AzureServiceErrorCode.PATH_NOT_FOUND.getErrorCode(),
+            "openFileForWrite must be used with files and not directories",
+            null);
       }
+      final Long contentLength = Long.valueOf(op.getResult()
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: We can use the extractContentLength() method proposed above here as well.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java [hunk 6]
```diff
@@ -127,8 +128,10 @@
 import static org.apache.hadoop.fs.azurebfs.constants.HttpHeaderConfigurations.IF_NONE_MATCH;
 import static org.apache.hadoop.fs.azurebfs.constants.HttpHeaderConfigurations.RANGE;
 import static org.apache.hadoop.fs.azurebfs.constants.HttpHeaderConfigurations.USER_AGENT;
+import static org.apache.hadoop.fs.azurebfs.constants.HttpHeaderConfigurations.X_MS_BLOB_CONTENT_MD5;
 import static org.apache.hadoop.fs.azurebfs.constants.HttpHeaderConfigurations.X_MS_BLOB_TYPE;
 import static org.apache.hadoop.fs.azurebfs.constants.HttpHeaderConfigurations.X_MS_COPY_SOURCE;
+import static org.apache.hadoop.fs.azurebfs.constants.HttpHeaderConfigurations.X_MS_EXISTING_RESOURCE_TYPE;
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java]: Nit; Unused Import

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java [hunk 7]
```diff
@@ -404,14 +408,9 @@ public AbfsRestOperation createPath(final String path,
...
     if (!isNamespaceEnabled && !isCreateCalledFromMarkers) {
-      Path parentPath = new Path(path).getParent();
-      if (parentPath != null && !parentPath.isRoot()) {
-        createMarkers(parentPath, overwrite, permissions, isAppendBlob, eTag,
-            contextEncryptionAdapter, tracingContext, isNamespaceEnabled);
-      }
       AbfsHttpOperation op1Result = null;
       try {
-        op1Result = getPathStatus(path, tracingContext,
+        op1Result = getPathStatus(path, new TracingContext(tracingContext),
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java]: Why do we need is implicit check as true here??
If we need to check for implicit then the Log message below needs to be changed.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java [hunk 8]
```diff
@@ -435,6 +434,11 @@ public AbfsRestOperation createPath(final String path,
               null);
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java:426]: These conditions can be combined as follows:
```
      if (op1Result != null) {
        boolean isDir = checkIsDir(op1Result);
        if (isFile == isDir) {
          throw new AbfsRestOperationException(HTTP_CONFLICT,
              AzureServiceErrorCode.PATH_CONFLICT.getErrorCode(),
              PATH_EXISTS,
              null);
        }
      }
```

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java [hunk 9]
```diff
@@ -435,6 +434,11 @@ public AbfsRestOperation createPath(final String path,
...
         }
       }
+      Path parentPath = new Path(path).getParent();
+      if (parentPath != null && !parentPath.isRoot()) {
+        createMarkers(parentPath, overwrite, permissions, isAppendBlob, eTag,
+            contextEncryptionAdapter, tracingContext, isNamespaceEnabled);
+      }
     }
     if (isFile) {
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java:435]: Nit: There is a condition few lines below which checks if its not a file. These 2 can be combined as if else construct.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java [hunk 10]
```diff
@@ -520,9 +531,10 @@ public void createMarkers(final Path path,
   private void checkParentChainForFile(Path path, TracingContext tracingContext,
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java]: Nit: createMarkers() method above should also be private.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java [hunk 11]
```diff
@@ -520,9 +531,10 @@ public void createMarkers(final Path path,
...
       List<Path> keysToCreateAsFolder) throws AzureBlobFileSystemException {
     AbfsHttpOperation opResult = null;
+    TracingContext tracingContext1 = new TracingContext(tracingContext);
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java]: This we already do in AbfsRestOperation before calling completeExecute method.
Was that not enough?

**Review by anujmodi2021 — CHANGES_REQUESTED** [review]
Reviewed production code changes...
Will review test code changes separately.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java (hunk 1 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java]: Has been changed in the new commit.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java (hunk 11 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java]: This was missing in databricks code, hence added here as well. Will revert here as lastTracingContext already does this

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAbfsNetworkStatistics.java [hunk 12]
```diff
@@ -89,7 +89,7 @@ public void testAbfsHttpSendStatistics() throws IOException {
        // 1 create request = 1 connection made and 1 send request
       if (client instanceof AbfsBlobClient && !getIsNamespaceEnabled(fs)) {
         expectedRequestsSent += (directory);
-        expectedConnectionsMade += (directory * 2);
+        expectedConnectionsMade += ((directory * 2) + 1);
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAbfsNetworkStatistics.java:93]: Nit: Please add comment why this change was made.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemAppend.java [hunk 13]
```diff
@@ -434,12 +463,19 @@ public void testCreateExplicitDirectoryOverDfsAppendOverBlob() throws IOExceptio
   public void testRecreateAppendAndFlush() throws IOException {
     final AzureBlobFileSystem fs = getFileSystem();
     final Path filePath = path(TEST_FILE_PATH);
+    Assume.assumeFalse("Not valid for APPEND BLOB",
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemAppend.java]: There was a method added in AbstractAbfsIntegrationTest for checking appendBlobEnable.
We can use that here and other places needed for cleaner code.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemCreate.java [hunk 14]
```diff
@@ -312,7 +318,7 @@ public void testCreateFileOverwrite(boolean enableConditionalCreateOverwrite)
     // 1. getFileStatus on DFS endpoint : 1
     //    getFileStatus on Blob endpoint: 2 (Additional List blob call)
     // 2. actual create call: 1
-    createRequestCount += (client instanceof AbfsBlobClient && !getIsNamespaceEnabled(fs) ? 2: 1);
+    createRequestCount += (client instanceof AbfsBlobClient && !getIsNamespaceEnabled(fs) ? 3: 1);
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemCreate.java:321]: Comments for changes made

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemCreate.java [hunk 15]
```diff
@@ -352,7 +358,7 @@ public void testCreateFileOverwrite(boolean enableConditionalCreateOverwrite)
     // 1. getFileStatus on DFS endpoint : 1
     //    getFileStatus on Blob endpoint: 2 (Additional List blob call for non-existing path)
     // 2. actual create call: 1
-    createRequestCount += (client instanceof AbfsBlobClient && !getIsNamespaceEnabled(fs) ? 2: 1);
+    createRequestCount += (client instanceof AbfsBlobClient && !getIsNamespaceEnabled(fs) ? 3: 1);
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemCreate.java:361]: Comments for changes made

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemCreate.java [hunk 16]
```diff
@@ -634,11 +685,113 @@ public void testCreatePathParentExplicit() throws Exception {
...
+
+  /**
+   * Test create on implicit directory with explicit parent.
+   * @throws Exception
+   */
+  @Test
+  public void testParentExplicitPathImplicit() throws Exception {
+    final AzureBlobFileSystem fs = getFileSystem();
+    final AzureBlobFileSystemStore store = fs.getAbfsStore();
+    Assume.assumeTrue(fs.getAbfsStore().getClientHandler().getIngressClient() instanceof AbfsBlobClient);
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemCreate.java:716]: We also have getAbfsServiceType() method in base class. That can be used to assert the blob endpoint.
`Assume.assumeTrue(getAbfsServiceType() == AbfsServiceType.BLOB)`

Here and other places it is used.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemCreate.java [hunk 17]
```diff
@@ -634,11 +685,113 @@ public void testCreatePathParentExplicit() throws Exception {
...
+    Path filePath = new Path(fileName);
+    fs.create(filePath);
+    String eTag = extractFileEtag(fileName);
+
+    // testing createFile on already existing file path
+    fs.create(filePath, true);
+
+    String eTagAfterCreateOverwrite = extractFileEtag(fileName);
+
+    assertFalse(
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemCreate.java]: Steve would prefer Assertions.assert()
We can change this when moving to OSS

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMkDir.java [hunk 18]
```diff
@@ -127,7 +131,7 @@ public void testCreateDirOverwrite(boolean enableConditionalCreateOverwrite)
     // One request to server
     AbfsClient client = fs.getAbfsStore().getClientHandler().getIngressClient();
     if (client instanceof AbfsBlobClient && !getIsNamespaceEnabled(fs)) {
-      mkdirRequestCount +=2;
+      mkdirRequestCount +=3;
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMkDir.java:135]: Comment for the change

**Review by anujmodi2021 — COMMENTED** [review]
Greate test coverage.
Only nits. Nothing major. 
Feel free to take Nits later.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java (hunk 2 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java]: Taken

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 3 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: Taken

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 4 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: Taken

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 5 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: Taken

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java (hunk 6 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java]: taken

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java (hunk 7 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java]: We should not create a file where a directory is already present be it implicit or explicit, hence we need implicit check as well. Updated the log

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java (hunk 8 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java:426]: taken

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java (hunk 9 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java:435]: taken

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java (hunk 10 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java]: taken

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAbfsNetworkStatistics.java (hunk 12 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAbfsNetworkStatistics.java:93]: taken

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemAppend.java (hunk 13 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemAppend.java]: taken

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemCreate.java (hunk 14 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemCreate.java:321]: already mentioned in comments above

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemCreate.java (hunk 16 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemCreate.java:716]: I want to check for the ingress client to be BLOB

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemCreate.java (hunk 17 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemCreate.java]: taken

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMkDir.java (hunk 18 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMkDir.java:135]: added

**Review by anujmodi2021 — APPROVED** [review]
+1 LGTM
